app/src/main/java/org/greenstand/android/TreeTracker/managers/Locations.kt

```diff
@@ -0,0 +1,332 @@
+     * method of computing variance but it calculates running cumulative variance but we need
+     * sliding window computation here.
+     *
+     * https://math.stackexchange.com/questions/2815732/calculating-standard-deviation-of-a-moving-window
+     *
+     * Assuming you are using SD with Bessel's correction, call μn and SDn the mean and
+     * standard deviation from n to n+99. Then, calculate μ1 and SD1 afterwards, you can use the
+     * recursive relation
+     *  μn+1=μn−(1/99*X(n))+(1/99*X(n+100)) and
+     *  variance(n+1)= variance(n) −1/99(Xn−μn)^2 + 1/99(X(n+100)−μ(n+1))^2
+     */
+    fun computeSlidingWindowStats(
+        currentStats: ConvergenceStats,
+        replacingValue: Double,
+        newValue: Double
+    ): ConvergenceStats {
+        val newMean = (
+                currentStats.mean -
+                        (replacingValue / CONVERGENCE_DATA_SIZE) +
+                        (newValue / CONVERGENCE_DATA_SIZE))
+        val newVariance = (
+                currentStats.variance -
+                        ((replacingValue - currentStats.mean).pow(2.0) / CONVERGENCE_DATA_SIZE) +
+                        ((newValue - newMean).pow(2.0) / CONVERGENCE_DATA_SIZE))
+        val newStdDev = sqrt(newVariance)
+        return ConvergenceStats(newMean, newVariance, newStdDev)
+    }
+
+    fun computeConvergence() {
+        Timber.d("Convergence: Evaluating initial convergence stats")
+        if (locations.size < CONVERGENCE_DATA_SIZE)
+            return
+        val longitudeData = locations.map { it.longitude }.toList()
+        longitudeConvergence = computeStats(longitudeData)
+
+        val latitudeData = locations.map { it.latitude }.toList()
+        latitudeConvergence = computeStats(latitudeData)
+    }
+
+    fun computeSlidingWindowConvergence(replaceLocation: Location, newLocation: Location) {
+        Timber.d("Convergence: Evaluating running convergence stats")
+        longitudeConvergence = computeSlidingWindowStats(
+            longitudeConvergence!!,
+            replaceLocation.longitude,
+            newLocation.longitude
+        )
+        latitudeConvergence = computeSlidingWindowStats(
+            latitudeConvergence!!,
+            replaceLocation.latitude,
+            newLocation.latitude
+        )
+    }
+
+    fun longitudinalStandardDeviation(): Double? {
+        return longitudeConvergence?.standardDeviation
+    }
+
+    fun latitudinalStandardDeviation(): Double? {
+        return latitudeConvergence?.standardDeviation
+    }
+}
+
+data class ConvergenceStats(
+    val mean: Double,
+    val variance: Double,
+    val standardDeviation: Double
+)
+
+data class LocationData(
+    val planterCheckInId: Long?,
+    val latitude: Double,
+    val longitude: Double,
+    val accuracy: Float,
+    val treeUuid: String?,
+    val capturedAt: Long
+)
+
+fun <T1 : Any, T2 : Any, R : Any> safeLet(p1: T1?, p2: T2?, block: (T1, T2) -> R?): R? {
```

Comment:
Let's see if we can remove this.

Reply:
I find this very useful and appropriate to do a null check on two variables. While this should be a utility function in a Util file,  it doesn't have multiple usages or other custom functions that we can group and move this out in its separate file yet.

Let me know if you have some other better way to replace it.

Reply:
I commented where this is used on a replacement.

This is something that we have to remember and each time it's seen it's another layer to understand and add complexity for other people.

Another way is to assign the two fields to local vals and do an if statement on them

val a = something?.thing()
val b = somethingelse?.thing()

if (a != null && b != null) {...}

Reply:
Sure, updating it with your suggestion.